File: FontendQuanLiChiTieu/src/services/user.service.ts
import api from './api';

export interface UserProfile {
    id: number;
    email: string;
    firstName: string;
    lastName: string;
    phone?: string;
    birthday?: string;
    avatarUrl?: string;
    currency: string;
    gender?: string;
}

export interface UpdateProfileRequest {
    firstName?: string;
    lastName?: string;
    phone?: string;
    birthday?: string;   
    gender?: string;    
    currency?: string;  
}

interface ApiResponse<T> {
    code: number;
    message?: string;
    result: T;
}

const UserService = {
    getMyProfile: async (): Promise<UserProfile> => {
        const response = await api.get<ApiResponse<UserProfile>>('/users/me');
        return response.data.result;
    },

    updateMyProfile: async (data: UpdateProfileRequest): Promise<UserProfile> => {
        const response = await api.put<ApiResponse<UserProfile>>('/users/me', data);
        return response.data.result;
    },

    uploadAvatar: async (asset: { uri: string; mimeType?: string }): Promise<string> => {
        const formData = new FormData();
        // @ts-ignore — same pattern as auth.service.ts register
        formData.append('file', {
            uri: asset.uri,
            name: 'avatar.jpg',
            type: asset.mimeType ?? 'image/jpeg',
        });
        const response = await api.post<ApiResponse<string>>('/users/me/avatar', formData, {
            headers: { 'Content-Type': 'multipart/form-data' },
        });
        return response.data.result;
    },
};

export default UserService;

File: AppQuanLiChiTieu/AppQuanLiChiTieu/src/main/java/com/example/AppQuanLiChiTieu/service/UserService.java
package com.example.AppQuanLiChiTieu.service;

import com.example.AppQuanLiChiTieu.dto.request.UpdateProfileRequest;
import com.example.AppQuanLiChiTieu.dto.response.UserProfileResponse;
import com.example.AppQuanLiChiTieu.model.User;
import com.example.AppQuanLiChiTieu.exception.AppException;
import com.example.AppQuanLiChiTieu.exception.ErrorCode;
import com.example.AppQuanLiChiTieu.repository.RedisUserRepository;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;

@Service
@RequiredArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
public class UserService {

    RedisUserRepository redisUserRepository;
    FileStorageService fileStorageService;
    
    // ── helpers ──────────────────────────────────────────────────────────────

    private User getCurrentUser() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return redisUserRepository.findById(email)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_EXISTED));
    }

    private UserProfileResponse toProfileResponse(User user) {
        return UserProfileResponse.builder()
                .id(user.getId())
                .email(user.getEmail())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .phone(user.getPhone())
                .birthday(user.getBirthday())
                .avatarUrl(user.getAvatarUrl())
                .currency(user.getCurrency())
                .gender(user.getGender())
                .isPremium(user.getIsPremium() != null && user.getIsPremium())
                .build();
    }

    // ── public API ───────────────────────────────────────────────────────────

    /** GET /users/me */
    public UserProfileResponse getMyProfile() {
        return toProfileResponse(getCurrentUser());
    }

    /** PUT /users/me */
    public UserProfileResponse updateMyProfile(UpdateProfileRequest request) {
        User user = getCurrentUser();
        if (request.getFirstName() != null) user.setFirstName(request.getFirstName());
        if (request.getLastName()  != null) user.setLastName(request.getLastName());
        if (request.getPhone()     != null) user.setPhone(request.getPhone());
        if (request.getBirthday()  != null) user.setBirthday(request.getBirthday());
        if (request.getGender()    != null) user.setGender(request.getGender());
        if (request.getCurrency()  != null) user.setCurrency(request.getCurrency());
        redisUserRepository.save(user);
        return toProfileResponse(user);
    }

    /** POST /users/me/avatar — upload and persist avatar image */
    public String uploadAvatar(MultipartFile file) throws IOException {
        if (file.isEmpty()) throw new AppException(ErrorCode.INVALID_KEY);

        String avatarUrl = fileStorageService.uploadImage(file);

        User user = getCurrentUser();
        user.setAvatarUrl(avatarUrl);
        redisUserRepository.save(user);

        return avatarUrl;
    }
}
